File: OneSignalSDK/onesignal/src/main/java/com/onesignal/language/LanguageContext.java
package com.onesignal.language;

import android.support.annotation.NonNull;
import android.util.Log;

import com.onesignal.OSSharedPreferences;

import static com.onesignal.language.LanguageProviderAppDefined.PREFS_OS_LANGUAGE;

/*
The Interface implements a getter and setter for the Language Provider.
It defaults to the device defined Language unless a language override is set.
 */
public class LanguageContext {
    private LanguageProvider strategy;
    private static LanguageContext instance = null;

    public static LanguageContext getInstance() {
        return instance;
    }

    public LanguageContext(OSSharedPreferences preferences) {
        instance = this;
        String languageAppDefined = preferences.getString(preferences.getPreferencesName(), PREFS_OS_LANGUAGE, null);
        if (languageAppDefined != null) {
            this.strategy = new LanguageProviderAppDefined(preferences);
        }
        else {
            this.strategy = new LanguageProviderDevice();
        }
    }

    public void setStrategy(LanguageProvider strategy) {
        this.strategy = strategy;
    }

    @NonNull
    public String getLanguage() {
        return strategy.getLanguage();
    }
}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/onesignal/core/internal/language/impl/LanguageContext.kt
package com.onesignal.onesignal.core.internal.language.impl

import com.onesignal.onesignal.core.internal.language.ILanguageProvider
import com.onesignal.onesignal.core.internal.preferences.IPreferencesService
import com.onesignal.onesignal.core.internal.preferences.PreferenceStores

/*
The Interface implements a getter and setter for the Language Provider.
It defaults to the device defined Language unless a language override is set.
 */
class LanguageContext(
    private val _preferences: IPreferencesService) {

    private var strategy: ILanguageProvider? = null

    init {
        instance = this
        val languageAppDefined = _preferences.getString(PreferenceStores.ONESIGNAL,
            LanguageProviderAppDefined.PREFS_OS_LANGUAGE, null)

        strategy = if (languageAppDefined != null) {
            LanguageProviderAppDefined(_preferences)
        } else {
            LanguageProviderDevice()
        }
    }

    fun setStrategy(strategy: ILanguageProvider?) {
        this.strategy = strategy
    }

    val language: String
        get() = strategy!!.language

    companion object {
        var instance: LanguageContext? = null
            private set
    }
}